tests/testsuite/path.rs

```diff
@@ -1919,3 +1918,144 @@ foo v1.0.0 ([ROOT]/foo)
 "#]])
         .run();
 }
+
+#[cargo_test]
+fn invalid_package_name_in_path() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name = "foo"
+            version = "0.5.0"
+            edition = "2015"
+            authors = []
+
+            [workspace]
+
+            [dependencies]
+            definitely_not_bar = { path = "crates/bar" }
+        "#,
+        )
+        .file("src/lib.rs", "")
+        .file(
+            "crates/bar/Cargo.toml",
+            r#"
+            [package]
+            name = "bar"
+            version = "0.5.0"
+            edition = "2015"
+            authors = []
+            "#,
+        )
+        .file("crates/bar/src/lib.rs", "")
+        .build();
+
+    p.cargo("check")
+        .with_status(101)
+        .with_stderr_data(str![[r#"
+error: no matching package named `definitely_not_bar` found at `bar/`
+note: required by package `foo v0.1.0 (/Users/eric/Temp/foo)`
+
+help: package `bar` exists at `bar/`
+"#]])
```

Comment:
As a Draft, I appreciate showing your intent in shopping this around.

As you work on this, I'd recommend making the tests pass, showing the current behavior.  Your next commit would then change Cargo's behavior and update the tests to show the new test output.